### PR TITLE
Make classes hashable

### PR DESCRIPTION
**Pull request by pala1212**

### CODE REVIEW

**Comment by adamjstewart** [comment]
Needs unit tests

File: fiscalyear.py [hunk 1]
```diff
@@ -226,10 +226,16 @@ def _check_quarter(quarter):
...
+class _Hashable(object):
+    def __hash__(self):
+        return hash(tuple(map(lambda x: getattr(self, x), self.__slots__)))
+
+
+class FiscalYear(_Hashable):
     """A class representing a single fiscal year."""
 
-    __slots__ = "_fiscal_year"
+    __slots__ = "_fiscal_year",
```
Comment (adamjstewart) [inline fiscalyear.py:237]:
```suggestion
    __slots__ = ["_fiscal_year"]
```
This is probably more clear and uniform

File: fiscalyear.py [hunk 2]
```diff
@@ -446,10 +452,11 @@ def __ge__(self, other):
             )
 
 
-class FiscalQuarter(object):
+class FiscalQuarter(_Hashable):
     """A class representing a single fiscal quarter."""
 
     __slots__ = ["_fiscal_year", "_fiscal_quarter"]
+    __hash__ = _Hashable.__hash__
```
Comment (adamjstewart) [inline fiscalyear.py:459]: Why do we need to reimplement `__hash__`?

**Comment by adamjstewart** [comment]
Can you run `black .` on the code?

File: fiscalyear.py (hunk 2 above)
Comment (pala1212) [reply fiscalyear.py:459]: Or else it wont inherit from  base class

Comment (adamjstewart) [reply fiscalyear.py:459]: Isn't the whole point of subclasses so that common attributes will be inherited?

Comment (pala1212) [reply fiscalyear.py:459]: From python docs:
A class that overrides __eq__() and does not define __hash__() will have its __hash__() implicitly set to None.
If a class that overrides __eq__() needs to retain the implementation of __hash__() from a parent class, the interpreter must be told this explicitly by setting __hash__ = <ParentClass>.__hash__

**Comment by pala1212** [comment]
> Can you run `black .` on the code?
Black does not fail for me....

(venv) gitpala@DESKTOP-8RRFQ6C:~/fiscalyear$ black --check .
All done! ✨ 🍰 ✨
4 files would be left unchanged.
(venv) gitpala@DESKTOP-8RRFQ6C:~/fiscalyear$ black --version
black, 21.12b0 (compiled: no)

File: fiscalyear.py (hunk 1 above)
Comment (pala1212) [reply fiscalyear.py]: https://github.com/adamjstewart/fiscalyear/pull/24/commits/ba817e239c99dbb56e98ede806f2045eee2a9eec

File: fiscalyear.py [hunk 3]
```diff
@@ -226,10 +226,16 @@ def _check_quarter(quarter):
         )
 
 
-class FiscalYear(object):
+class _Hashable(object):
+    def __hash__(self):
```
Comment (adamjstewart) [inline fiscalyear.py:230]: Can you add a docstring? Something like:
```python
"""Unique hash of an object instance.

:returns: a unique hash
:rtype: int
"""
```

**Review by adamjstewart — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**